### PR TITLE
Update Starscream to 4.0.6

### PR DESCRIPTION
**Pull request by crazytonyli**
## Description

This PR updates starscream to the latest version, so that we can update it later to be compliant with [Apple's third-party SDK requirements](https://href.li/?https://developer.apple.com/support/third-party-SDK-requirements).

I have to change the existing `PinghubClient` code to adopt the breaking changes in Starscream.

I have added new unit tests to test the websocket connection from `PinghubClient` to a websocket server, by starting a real websocket server for the unit tests.

Initially, I tried to implement a websocket server using the `Network` framework directly. But I couldn't get it to work. I switched to the [`WebsocketServer`](https://github.com/daltoniam/Starscream/blob/4.0.6/Sources/Server/WebSocketServer.swift) came with Starscream.

### Potential next steps

Starscream v3 is built on top of [socket API](https://github.com/daltoniam/Starscream/blob/3.1.2/Sources/Starscream/WebSocket.swift). v4 adds a couple of alternatives on top of the v3 implementation: one using [`URLSession.webSocketTask(with)`](https://developer.apple.com/documentation/foundation/urlsession/3181171-websockettask), and the other using `Network` framework. [By default, the `Network` framework one is used](https://github.com/daltoniam/Starscream/blob/4.0.6/Sources/Starscream/WebSocket.swift#L119).

IMO, we don't _need_ a third-party implementation, we can just use [`URLSession.webSocketTask(with)`](https://developer.apple.com/documentation/foundation/urlsession/3181171-websockettask) directly. In fact, if we tells Starscream [not to use its own implementation](https://github.com/daltoniam/Starscream/blob/4.0.6/Sources/Starscream/WebSocket.swift#L121), it becomes [a simple wrapper around `URLSession.webSocketTask`](https://github.com/daltoniam/Starscream/blob/4.0.6/Sources/Engine/NativeEngine.swift).

So, I think it's worth considering removing Starscream from the app. The refactor would become slightly easier, now that there are new unit tests in this PR which use real websocket connections.

## Test Instructions

1. Launch Jetpack app and sign in using a WP.com account.
2. Go to the Notifications tab.
3. Use another WP.com account to comment/like posts on the signed-in account's site.
4. Send the app to background and bring it back. Repeat the above step.

Upon comment/like actions, you should see logs like "PingHub push, syncing note" in Xcode console and the Notification tab will be updated in real-time.

## Regression Notes
1. Potential unintended areas of impact
None.

5. What I did to test those areas of impact (or what existing automated tests I relied on)
What's described in the Test Instructions section

6. What automated tests I added (or what prevented me from doing so)
Unit tests.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist: N/A

### CODE REVIEW

File: WordPress/Classes/Networking/Pinghub.swift [hunk 1]
```diff
@@ -218,13 +221,85 @@ internal protocol Socket: AnyObject {
 // MARK: - Starscream
 
 private func starscreamSocket(url: URL, token: String) -> Socket {
-    var request = URLRequest(url: PinghubClient.endpoint)
+    var request = URLRequest(url: url)
     request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
-    return WebSocket(request: request)
+    return StarscreamWebSocket(socket: WebSocket(request: request))
```
Comment (crazytonyli) [inline WordPress/Classes/Networking/Pinghub.swift:226]: By default, this `WebSocket` initialiser's `useCustomEngine` parameter is true. We could change it to be`WebSocket(request: request, useCustomEngine: false)` which leads Starscream to use `URLSession.webSocket(with:)`. The app works out exactly the same based on my testing.

However, that change causes the new unit tests to fail. I suspect it's an implementation issue in Starscream's `WebSocketServer`. I didn't look into it further, thus didn't change this initializer.

Comment (mokagio) [reply WordPress/Classes/Networking/Pinghub.swift:226]: I guess whether to look more into this will depend on whether we decide to remove Starscream.

**Review by mokagio — APPROVED** [review]
> IMO, we don't need a third-party implementation, we can just use URLSession.webSocketTask(with) directly. In fact, if we tells Starscream not to use its own implementation, it becomes a simple wrapper around URLSession.webSocketTask.
>
> So, I think it's worth considering removing Starscream from the app. The refactor would become slightly easier, now that there are new unit tests in this PR which use real websocket connections.

+1 on the idea of dropping Starscream 👍

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/271">24.3</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Pull request merged**